Rclone Tutorial (#67)
hetzneronline/community-content: tutorials/rclone-basics/01.en.md

### Step 3.1 - &lt;Moving Files Using Rclone>
Some example commands used to explore folder and files in Rclone:
```
    rclone lsd [remote name]:/[Optional - Specify Remote Folder]
    (List the directories of the top level folder or specific folder)

    rclone ls [remote name]:/[Optional - Specify Remote Folder]
    (List all the files in the remote or specific folder)

    rclone lsl [remote name]:/[Optional - Specify Remote Folder]
    (List location / size / of all files in remote or specific folder)
```

We can see the output of the ```ls``` below.

![ls Viewing](https://github.com/sharkeydavid/community-content/blob/master/tutorials/rclone-basics/images/ls-file.png)

*Getting Fill Files List*

All the commands are listed in the Docs [here](https://rclone.org/docs/#subcommands).

Note above 2 things:

* Running the **ls** command on the whole remote will produce a long list of files, it’s better to specify a specific folder and run the command there.

* If your folder has spaces in its name, use quotation markets to bind it together. Example: Temp folder → ‘Temp Folder’

### Step 3.2 - &lt;Moving Files Using Rclone>

> Moving files and folders is easy once you understand the structure. I have set up dedicated servers on [Hetzner](https://www.hetzner.com/) where once SSH’d in, I manage and move 800GB at a time using the Gigabit speeds to transfer files between services. Critical files to clients, family, or friends can be moved quickly and without downloading them to your local machine.

Author: Removed

Below are the steps required to transfer files between 2 providers, GDrive and Dropbox. Remember to be careful when using some of these commands as they can be destructive (Files get deleted or removed). I can’t help if you move files and they get lost or deleted.

The most basic command is:

```
    rclone copy [Source]:/[Folder] [Destination)]:/[Folder]

    **rclone copy** <-- This is the copy command that Copy files from source to dest, skipping already copied. Use this one for first sync as files won’t be deleted when used.

    **[Source]:/[Folder]** <-- Source is the remote containing the files. And if you are only syncing a specific folder, then ensure you specify this folder or you will sync the whole remote.

    **[Destination)]:/[Folder]** <-- Destination is the remote recieving the files. And if you are putting the files in specific folder, then ensure you specify this folder or you end up with a messy remote.

> ```

Author: i dont, good point. I have put only the commands in a code block.

![Moving Files](https://github.com/sharkeydavid/community-content/blob/master/tutorials/rclone-basics/images/sample-move.png)

*Moving Files Between Remotes*

You can see that the files are mirrored from my GDrive to my Dropbox.

Note: A bonus to using Rclone is that file metadata is preserved. If a file is dated in the past on Remote1, the same date is seen on Remote2. Very handy when transferring large sets of files.

Additionally, we could use a command ```sync```, to move files between remotes.

Note: **Destructive** use of SYNC. If we want to exactly mirror 2 remotes, this can be very useful. If we want to backup 2 remotes this is not useful as **all changes are copied from the *source* remote to the *destination* remote, including any files deleted will be deleted in the destination remote**.

### Step 4 - &lt;Bonus - Flags>

> Getting fine control of Rclone can unlock many benifits such as increased speed, updating, and more. Below are some of my favourites broken down:

Reviewer: ```suggestion
Getting fine control on Rclone can unlock many benefits such as increased speed, updating, and more. Below are some of my favorites broken down:
```

```rclone copy [Source]: [Dest]: -v -c --bwlimit 8M --transfers=3 --stats=30s --dry-run```

> ```-v``` <-- Using the -v flag, rclone will produce Error, Notice and Info messages.

Reviewer: ```suggestion
`-v` <-- Using the -v flag, rclone will produce Error, Notice and Info messages.
```

> ```-c``` <-- Rclone will look at modification time and size of files to see if they are equal. If you set this flag then rclone will check the file hash and size to determine if files are equal.

Reviewer: ```suggestion
`-c` <-- Rclone will look at modification time and size of files to see if they are equal. If you set this flag then rclone will check the file hash and size to determine if files are equal.
```

> ```--bwlimit 8M``` <-- Taking up all the bandwidth is rude. I fif you need to limit the speed of the transfers use this flag. Example: GDrive limits daily transfers to 750GB, or ~9MB/s.

Reviewer: ```suggestion
`--bwlimit 8M` <-- Taking up all the bandwidth is rude. If you need to limit the speed of the transfers use this flag. Example: Google Drive limits daily transfers to 750GB, or ~9MB/s.
```

> ```--transfers=3``` <-- Specify the number of files being transferred in parallal. The default is 4, but I have gone up to 20 without any issues.

Reviewer: ```suggestion
`--transfers=3` <-- Specify the number of files being transferred parallel. The default is 4, but I have gone up to 20 without any issues.
```

> ```--stats=30s``` <-- Rclone will produce an update stat every X seconds or minutes to show you the rate or transfer or other key details.

Reviewer: ```suggestion
`--stats=30s` <-- Rclone will produce an update for stats every X seconds or minutes to show you the rate or transfer and other key details.
```

> ```--dry-run``` <-- Before commiting to a transfer and potentially regretting it, run the command with this flag. It will show you what will happen when you run a copy/sync/etc without performing it. One you are confident, remove this flag and repeat the command.

Reviewer: ```suggestion
`--dry-run` <-- Before committing to a transfer and potentially regretting it, run the command with this flag. It will show you what will happen when you run a copy/sync/etc without performing it. Once you are confident, remove this flag and repeat the command.
```

Some more useful flags:

Update Remotes:

> ```--update``` <-- Rclone will be forced to skip any files which exist on the destination and have a modified time that is newer than the source file.

Reviewer: ```suggestion
`--update` <-- Rclone will be forced to skip any files which exist on the destination and have a modified time that is newer than the source file.
```

Progress Block:

> ```--progress``` <-- Rclone update the stats in a static block in the terminal providing a realtime overview of the transfer. This would be used in place of --stats.

Reviewer: ```suggestion
`--progress` <-- Rclone updates the stats in a static block in the terminal providing a real-time overview of the transfer. This would be used in place of `--stats`.
```

Immutable

> ```--immutable``` <-- Rclone treat the source and destination files as immutable and won't allow any modification. This does NOT stop files being deleted, existing files will only never be updated.

Reviewer: ```suggestion
`--immutable` <-- Rclone treat the source and destination files as immutable and won't allow any modification. This does NOT stop files being deleted, existing files will only never be updated.
```

### Terminology
* Remote: Online file storage system used to host files.

## Conclusion

> Rclone is a powerful tool. Whem combined with Hetzner Gigabit speeds, it can be used to manage huge quantities of information quickly and efficiently. I always advise on testing out moving files first, before attempting large data migrations. To ensure the commands and flags are as you expect. To do this, I personally have a small CX11 VPS. When moving large sets of files, i scale up the instance to 2-4 vCPUs and begin the move.

Reviewer: ```suggestion
Rclone is a powerful tool. When combined with Hetzner Gigabit speeds, it can be used to manage huge quantities of information quickly and efficiently. I always advise on testing out moving files first, before attempting large data migrations to ensure the commands and flags are as expect. In order to test this, I have a small CX11. When moving large sets of files, i scale up the instance to 2-4 vCPUs and begin the transfer.
```

> P.S: To keep files moving while not SSH'd in the server. A great feature of Linux is: ```screen```. Details on this can be found [here](https://linuxize.com/post/how-to-use-linux-screen/)

Reviewer: ```suggestion
P.S.: To keep files moving while not connected via SSH on the server. A great program for Linux is: `screen`. Details on this can be found [here](https://linuxize.com/post/how-to-use-linux-screen/).
```
